### PR TITLE
Phase 2

### PR DESCRIPTION
**Pull request by Kulkodar**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Game of Life
+
+The Game of Life algorithm implemented as a PHP CLI program.
+
+## Usage
+
+Run gol.php to run the program with its default parameter.
```
Comment (Shyru) [inline README.md:7]: s/parameter/parameters

File: src/rule.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace GOL;
+
+/**
+ * Represents the rule of the Game of Life.
+ *
+ * Use applyRule() to check if a cell lives or dies in the next generation.
+ */
+class Rule
```
Comment (Shyru) [inline src/rule.php:10]: Although it may hurt, please remove this class.

File: src/boards/board.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    protected $width;
+    protected $height;
+    protected $rule;
+
+    /**
+     * @param $_width int Width of the Board with margin.
+     * @param $_height int Height af the Board with margin.
+     * @param Rule $_rule rule used to generate the next generation.
+     */
+    public function __construct($_width, $_height, Rule $_rule)
```
Comment (Shyru) [inline src/boards/board.php:24]: Please remove rule-support for now. This is something that is a topic for a later phase.

File: src/boards/board.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace GOL\Boards;
+
+use GOL\Rule;
+
+/**
+ * Represents a Game of Life world.
+ *
+ * Use print() to print the board and nextGeneration() to calculate the next generation.
```
Comment (Shyru) [inline src/boards/board.php:10]: Good documentation!

File: src/boards/board.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                $buffer[$x][$y] = $this->rule->applyRule($this->countLivingNeighbours($x, $y), $this->grid[$x][$y]);
+
+        $this->grid = $buffer;
+    }
+
+    /**
+     * Returns the board.
+     * @return array 2d array with width and height of the board.
+     */
+    public function grid()
```
Comment (Shyru) [inline src/boards/board.php:71]: Why is this method necessary?

File: src/boards/board.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    {
+        return $this->grid;
+    }
+
+    /**
+     * Returns the width of the board including the margin.
```
Comment (Shyru) [inline src/boards/board.php:77]: What is the margin in terms of Game-of-Life?

File: src/boards/board.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+     * @return int Width of the board.
+     */
+    public function width()
+    {
+        return $this->width;
+    }
+
+    /**
+     * Returns the size of the board.
+     * @return int Size of the board.
```
Comment (Shyru) [inline src/boards/board.php:87]: s/size/height

File: src/boards/board.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    /**
+     * Returns the amount of living cells in the neighbourhood of a specific cell.
+     *
+     * No out of bound check due to the margin.
+     *
+     * @param $_x int y coordinate of the specific cell
+     * @param $_y int y coordinate of the specific cell
+     * @return int amount of living cells and -1 if given cell is out of bounds or on the margin.
+     */
+    public function countLivingNeighbours($_x, $_y)
```
Comment (Shyru) [inline src/boards/board.php:103]: Why is this method public? Does this need to be called from the outside? Couldn't this be `private` instead?

File: src/boards/boardacorn.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace GOL\Boards;
+
+use GOL\Rule;
+
+/**
+ * Represents a Game of Life world filled with an acorn in the center.
+ *
+ * Use grid() to retrieve the current grid and setGrid() to apply all changes.
```
Comment (Shyru) [inline src/boards/boardacorn.php:10]: Why is this important for the BoardAcorn class? (By the way I would prefer `AcornBoard` as name, but it is up to you to decide on that)

File: src/boards/history.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace GOL\Boards;
+
+/**
+ * Tracks previous version of a board.
+ *
+ * Use compare() to compare the current board with the previous versions.
+ * Use push() to add the current board to the history and pop() to remove the oldest board.
+ */
+class history
```
Comment (Shyru) [inline src/boards/history.php:11]: I like this class a lot! Good practice to put that into its own class!

**Review by Shyru — CHANGES_REQUESTED** [review]
Good work so far! :+1: Good documentation overall. You will need to refactor the specific board classes in the future, but for now we can leave it like this, I like it!

File: src/boards/board.php (hunk 6 above)
Comment (Kulkodar) [reply src/boards/board.php:77]: I made the array one element larger in every direction to avoid out of bounds check in the countLivingNeighbours method

**Pull request merged**